### ISSUE TITLE
Problems retrieving demo data

### ISSUE THREAD

**Issue by sespesogil** [thomasgilgenast/hic3defdr#1]
Hi! 

Congrats for the nice tool. I have problems to download the demo data. It seems hosted in a dropbox link that might be broken? Not 100% sure.  It would be great if you could share it with me. 

Thanks a lot in advance. 

Kind regards. 

S.

**Comment by thomasgilgenast** [comment]
Thanks for trying out the tool!

You are right, the demo data is not working because the Dropbox link it tries to download from is broken - sorry that you ran into this.

I will look into updating the demo data module to make it download from a different source.

As a short term workaround, you can download the demo data zip from [this Google Drive link](https://drive.google.com/file/d/11UVj3UUHbLXQuPAn3bTG-ZvBAwmvp4HY/view?usp=sharing) and then unzip that to `~/hic3defdr-demo-data` (so that a file like `~/hic3defdr-demo-data/ES_1/chr18_raw.npz` exists for example), then you should be able to run the rest of the demo as normal.

**Comment by sespesogil** [comment]
Thanks Thomas. Really appreciate it. 

Kind regards

S.

**Comment by sespesogil** [comment]
Dear Thomas, 

Thanks for sharing the demo data.  Just trying to reproduce the inputs for my data.  I would like to use my HiCPro and Mustache loops results if possible. Any suggestion how to produce the inputs directly from those tools?

Thanks a lot. Kind regards, 

S.

**Comment by thomasgilgenast** [comment]
This is a good question - compatibility with popular HiC tools like HiCPro and Mustache is definitely something we should include in hic3defdr, but we haven't implemented it there yet.

I can try to give a more detailed answer or some example code later this week when I have time, but to give a brief high-level overview of what I would try:
 - For HiCPro:
   - I would try to load the .matrix files into a separate sparse COO matrix for each chromosome (`scipy.sparse.coo_matrix()`; .matrix files are conceptually very similar to the `(row, column, value)` format of COO; note that you will need to use the HiCPro .bed file to figure out which HiCPro bin IDs belong to which chromosomes), then save them to disk with `scipy.sparse.save_npz()`. If you are not familiar with `scipy.sparse` this might be a bit tricky, so I will try to post an example later.
   - An alternative is to use HiCPro's [`hicpro2juicebox.sh`](https://github.com/nservant/HiC-Pro/blob/master/bin/utils/hicpro2juicebox.sh) to convert to .hic, get the sparse matrix data into Python via [straw](https://github.com/aidenlab/straw), and then create the sparse COO matrix and save to disk (with `scipy.sparse.coo_matrix()` and `scipy.sparse.save_npz()`).
 - For Mustache: the loop cluster format expected by hic3defdr is unfortunately a bit complicated, but I've tried to explain it a bit [here](https://hic3defdr.readthedocs.io/en/stable/overview/#sparse-json-cluster-format). If your contact matrix resolution is 10 kb and your Mustache output looks like this:
   ```
   chr1 <tab> 20000 <tab> 40000 <tab> chr1 <tab> 80000 <tab> 11000 <tab> ...
   chr1 <tab> 120000 <tab> 140000 <tab> chr1 <tab> 180000 <tab> 111000 <tab> ...
   ```
   then the hic3defdr clusters.json for chr1 would look like:
   ```
   {
     [[2, 8], [2, 9], [2, 10], [3, 8], [3, 9], [3, 10]],
     [[12, 18], [12, 19], [12, 110], [13, 18], [13, 19], [13, 110]],
   }
   ```
   Each line in the JSON represents one loop and lists the pixels (bin-bin pairs in bin units, not bp units) that are part of that loop.

**Comment by sespesogil** [comment]
Thanks Thomas, really appreciate it. 

I am gonna try. 

S.

**Comment by sespesogil** [comment]
Hi Thomas, 

I think it might not be that complicated to convert Mustache loop into json format. But in your example, the resulting json, 110 shouldn't be something different?  For example, correct me if I am wrong:

Given these loops:

chr1 <tab> 20000 <tab> 40000 <tab> chr1 <tab> 80000 <tab> 11000 <tab> ...
chr1 <tab> 120000 <tab> 140000 <tab> chr1 <tab> 180000 <tab> 200000 <tab> ...

Should end being:

{
  [[2, 8], [2, 9], [2, 10], [3, 8], [3, 9], [3, 10]],
  [[12, 18], [12, 19], [12, 20], [13, 18], [13, 19], [13, 20]],
}

Thanks!

**Comment by thomasgilgenast** [comment]
Good catch, my example was definitely wrong there - sorry, I'm not sure what I was thinking there!

I think your version is almost correct: when the second bed interval in the mustache output line is `chr1 180000 200000` then the y-values of the "pixels" that belong to the loop range from 180000/10000 = 18 to 200000/10000 = 20, but the right endpoint of that range is not included.

This is because when hic3defdr says that pixel `[13, 20]` is part of the loop, that corresponds to an x-axis bed interval of `chr1:130000-140000` and a y-axis bed interval of `chr1:200000-210000` - one pixel is one 10 kb by 10 kb square in bed interval terms. So when the mustache bed interval ends on 200000, the pixels whose y-axis bed intervals end at 200000 are the ones whose "index" is 19.

So I think the json should be:

```
{
  [[2, 8], [2, 9], [2, 10], [3, 8], [3, 9], [3, 10]],
  [[12, 18], [12, 19], [13, 18], [13, 19]],
}
```

Hopefully this makes sense, since `chr1 120000 140000 chr1 180000 200000` is a 20 kb by 20 kb bed interval square, so in terms of 10 kb pixels it should be 2 by 2 pixels (4 pixels total) in size.

**Comment by sespesogil** [comment]
Gotcha! Thanks a lot!

**Comment by sespesogil** [comment]
Hi Thomas again. Sorry for all questions. What about the KR.biases? What did you use to get those? 

Thanks!

S.

**Comment by thomasgilgenast** [comment]
This is a good question - in our internal pipeline the demo data files (including the KR bias vectors) were generated by a different tool (i.e., not by hic3defdr), so this part is not really explained in the tutorial.

However, there is a part of the tutorial that shows how the KR bias vectors can be computed from raw scipy sparse matrices stored on disk in .npz format: see the first code block in [this section](https://hic3defdr.readthedocs.io/en/stable/overview/#evaluating-simulations) for details:

```
import sys

import numpy as np
import scipy.sparse as sparse

from hic3defdr.util.filtering import filter_sparse_rows_count
from hic3defdr.util.balancing import kr_balance
from hic3defdr.util.printing import eprint


infile_pattern = 'sim/<rep>_<chrom>_raw.npz'
repnames = ['A1', 'A2', 'B1', 'B2']
chroms = ['chr18', 'chr19']

for repname in repnames:
    for chrom in chroms:
        eprint('balancing rep %s chrom %s' % (repname, chrom))
        infile = infile_pattern.replace('<rep>', repname)\
            .replace('<chrom>', chrom)
        outfile = infile.replace('_raw.npz', '_kr.bias')
        _, bias, _ = kr_balance(
            filter_sparse_rows_count(sparse.load_npz(infile)), fl=0)
        np.savetxt(outfile, bias)
```

The code block loops through a list of replicates and chromosomes and balances each raw matrix with the KR algorithm via the `kr_balance()` function. Before balancing it filters out sparse rows (which tend to impair convergence) with the `filter_sparse_rows_count()` function. In the tutorial, the raw matrices being balanced are simulated raw contact matrices generated by hic3defdr, but you could modify this code block (for example by just changing the `infile_pattern`, `repnames`, and `chroms`) to balance your own raw contact matrices.
